KAFKA-8037: Check deserialization at global state store restoration

pkleindl opened the pull request:
This change aims to prevent records that cannot be deserialized from going into a global state store during restore. Currently such records are filtered during normal operations but will be processed during restore and will cause an exception when trying to access the value in the store.
The change copies the logic from the GlobalStateUpdateTask and builds a list of deserializers to use during restoration.

The GlobalStateManagerImplTest was extended to cover the case that a StreamsException is expected when a record is processed that can't be deserialized with the default LogAndFailExceptionHandler.
GlobalStateManagerImplLogAndContinueTest was added and contains one new test which uses the LogAndContinueExceptionHandler and verifies that a record can be ignored during restoration.
Copying all tests is not ideal, but I found no easy way to override the DefaultExceptionHandler just for the one case.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Deserialize records on global state store restoration to prevent malformed records in store

pkleindl committed:
KAFKA-8037: Refactored unit tests

pkleindl committed:
KAFKA-8037: Refactored unit tests

pkleindl committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-8037

pkleindl commented:
@mjsax @vvcephei @guozhangwang May I kindly ask for a short review of my proposed changes for the global state stores?
If they are ok I can start adapting them for the local state stores too.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImplTest.java:
```diff
@@ -71,25 +79,28 @@
-    private File checkpointFile;
-    private ProcessorTopology topology;
-    private InternalMockProcessorContext processorContext;
+    final MockTime time = new MockTime();
```
instead of `package private` could these variables get `proctected` access?

pkleindl replied:
@bbejeck fixed and pushed, thanks for the feedback

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImplLogAndContinueTest.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Test
+    public void shouldThrowStreamsExceptionWhenRestoringWithLogAndFailExceptionHandler() {
```
The `GlobalStateManagerImplTest` already defines this test as well should this get removed here?

pkleindl replied:
@bbejeck This prevents the test from being executed, because it would fail because the exception would not be thrown in this setup. I could have prevented this by using the same name for the test but I thought a descriptive naming was more important.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImplTest.java:
```diff
@@ -117,20 +139,52 @@ public void before() {
+        consumer.addRecord(new ConsumerRecord<>(t5.topic(), t5.partition(), 3, longValue, longValue));
+
+        stateManager.initialize();
+        stateManager.register(store5, stateRestoreCallback);
```
Somewhat of nitpick but since this test is expected to fail maybe we can get rid of line 186.

Also, we've been moving away from declaring expected exceptions and using a form like 

```java
try {
      stateManager.register(store5, stateRestoreCallback);
      fail("Should not get here as LogAndFailExceptionHandler used");
   } catch (StreamsException e) {
            //expected ok to ignore
  }
```
Might be subjective on my part but the logic is little easier to follow

pkleindl replied:
@bbejeck fixed and pushed, thanks for the feedback

bbejeck reviewed (commented):
Thanks for the contributing @pkleindl.  Overall this looks good, I've left a few comments.

bbejeck commented:
ping @guozhangwang @mjsax @vvcephei @ableegoldman

bbejeck commented:
>GlobalStateManagerImplLogAndContinueTest was added and contains one new test which uses the LogAndContinueExceptionHandler and verifies that a record can be ignored during restoration.
Copying all tests is not ideal, but I found no easy way to override the DefaultExceptionHandler just for the one case.

I'm ok with this approach, we'll see what others have to say though.

pkleindl committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-8037

pkleindl committed:
KAFKA-8037: changed variables to protected, changed test to catch exception instead of using expected

pkleindl committed:
KAFKA-8037: changed variables to protected, changed test to catch exception instead of using expected

khaireddine120 committed:
KAFKA-7243: Add unit integration tests to validate metrics in Kafka Streams (#6080)

The goal of this task is to implement an integration test for the kafka stream metrics.

We have to check 2 things:
1. After streams application are started, all metrics from different levels (thread, task, processor, store, cache) are correctly created and displaying recorded values.
2. When streams application are shutdown, all metrics are correctly de-registered and removed.

Reviewers: John Roesler <john@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

mjsax committed:
KAFKA-8142: Fix NPE for nulls in Headers (#6484)

Reviewers: Bill Bejeck <bill@confluent.io>, Guozhang Wang <guozhang@confluent.io>

radai-rosenblatt committed:
MINOR: Avoid unnecessary collection copy in MetadataCache (#6397)

`map` was being used to convert `Iterable[Integer]` to `Iterable[Int`]. That
operation represented 11% of total CPU time measured under load for us.
We also expect a positive impact on GC.

Reviewers: Joel Koshy <jjkoshy@gmail.com>, Ismael Juma <ismael@juma.me.uk>

bbejeck committed:
MINOR: Clean up ThreadCacheTest (#6485)

Minor clean up ofThreadCacheTest
Reviewers: Guozhang Wang <wangguoz@gmail.com>, Matthias J. Sax <mjsax@apache.org>

brianbushree committed:
MINOR: list-topics should not require topic param

`kafka.list_topics(...)` should not require a topic parameter

Author: Brian Bushree <bbushree@confluent.io>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6367 from brianbushree/list-topics-no-topic

cmccabe committed:
MINOR: fix message protocol help text for ElectPreferredLeadersResult (#6479)

Reviewers: Jun Rao <junrao@gmail.com>

mjsax committed:
MINOR: add MacOS requirement to Streams docs

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

Author: Matthias J. Sax <matthias@confluent.io>

Reviewers: Bill Bejeck

Closes #6490 from mjsax/minor-streams-docs-rocksdb

bbejeck committed:
MINOR: Remove line for testing repartition topic name (#6488)

With KIP-307 joined.name() is deprecated plus we don't need to test for repartition topic names.
Reviewers: Matthias J. Sax <mjsax@apache.org>

kkonstantine committed:
KAFKA-8014: Extend Connect integration tests to add and remove workers dynamically (#6342)

Extend Connect's integration test framework to add or remove workers to EmbeddedConnectCluster, and choosing whether to fail the test on ungraceful service shutdown. Also added more JavaDoc and other minor improvements. 

Author: Konstantine Karantasis <konstantine@confluent.io>

Reviewers: Arjun Satish <arjun@confluent.io>, Randall Hauch <rhauch@gmail.com>

Closes #6342 from kkonstantine/KAFKA-8014

cmccabe committed:
KAFKA-8150: Fix bugs in handling null arrays in generated RPC code (#6489)

ToString functions must not get a NullPointException.  read() functions
must properly translate a negative array length to a null field.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

pierDipi committed:
MINOR: Fix misspelling in protocol documentation

Reviewers: Colin P. McCabe <cmccabe@apache.org>

stanislavkozlovski committed:
KAFKA-8102: Add an interval-based Trogdor transaction generator (#6444)

This patch adds a TimeIntervalTransactionsGenerator class which enables the Trogdor ProduceBench worker to commit transactions based on a configurable millisecond time interval.

Also, we now handle 409 create task responses in the coordinator command-line client by printing a more informative message

Reviewers: Colin P. McCabe <cmccabe@apache.org>

ivanyu committed:
KAFKA-7986: Distinguish logging from different ZooKeeperClient instances (#6493)

A broken can have more than one instance of ZooKeeperClient. For example, SimpleAclAuthorizer creates a separate ZooKeeperClient instance when configured.

This commit makes it possible to optionally specify the name for the ZooKeeperClient instance. The name is specified only for a broker's ZooKeeperClient instances, but not for commands' and tests'.

Reviewers: Jun Rao <junrao@gmail.com>

mjsax committed:
MINOR: Add 2.2.0 upgrade instructions (#6501)

Reviewers: Jason Gustafson <jason@confluent.io>

mjsax committed:
KAFKA-3522: Add RocksDBTimestampedSegmentedBytesStore (#6186)

Reviewers: Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Guozhang Wang <guozhang@confluent.io>

soenkeliebau committed:
MINOR: Fix some spelling corrections in comments (#6507)

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

dongjinleekr committed:
KAFKA-7502: Cleanup KTable materialization logic in a single place (filter) (#6453)

This PR is a follow-up of #6174, which handles doFilter / doMapValues / doTransformValues methods.

Reviewers: Bill Bejeck <bill@confluent.io>, Guozhang Wang <guozhang@confluent.io>

cmccabe committed:
MINOR: WorkerUtils#abort: fix bug in abort logic (#6516)

doneFuture is supposed to be completed with an empty string (meaning success) or a non-empty string which is the error message.  Currently, due to exception.getMessage sometimes returning null or an empty string, this is not working correctly.  This patch fixes that.

Reviewers: David Arthur <mumrah@gmail.com>

Tcheutchoua-Steve committed:
MINOR: Streams input topic corrected (#6513)

Though out the tutorial, the name of the input topic that was created is `streams-plaintext-input`. However, this was mistaken at some point in the tutorial and changed to `streams-wordcount-input`.

This patch is to adjust that. Thanks.

Reviewers: Guozhang Wang <wangguoz@gmail.com>

dongjinleekr committed:
Cleanup KTableImpl#doTransformValues (#6519)

This PR is a follow-up of #6174 and #6453, which cleans up KTableImpl#doTransformValues method.

Reviewers: Bill Bejeck <bbejeck@gmail.com>

dongjinleekr committed:
KAFKA-7502: Cleanup KTable materialization logic in a single place (doMapValues) (#6520)

* Move materialization logic from TableProcessorNode to KTableImpl

1. TableProcessorNode: remove materializedInternal, use storeBuilder instead.
2. Instantiate StoreBuilder in KTableImpl#[doFilter, doMapValues, doTransformValues], instead of TableProcessorNode#init.

* Cleanup KTableImpl#doMapValues

* 1. Add TableProcessorNode(String, ProcessorParameters, StoreBuilder). 2. Reformat+trivial changes on TableProcessorNode.java.

bbejeck committed:
MINOR: Move KTable source topic for changelog to optimization framework (#6500)

Since we've added Kafka Streams optimizations in 2.1 we need to move the optimization for source KTable nodes (use source topic as changelog) to the optimization framework.

Reviewers: Guozhang Wang <wangguoz@gmail.com>

mimaison committed:
KAFKA-8034: Use automatic RPC generation in DeleteTopics

Reviewers: Colin P. McCabe <cmccabe@apache.org>

cmccabe committed:
MINOR: Optimize ConnectionStressWorker

Optimize ConnectionStressWorker by avoiding creating a new
ChannelBuilder each time we want to open a new connection.

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Gwen Shapira

Closes #6518 from cmccabe/optimize-connection-stress-worker

mgrubent committed:
MINOR: Comment spelling nit

`CYGINW` probably should be `CYGWIN`

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

Author: Michael Gruben Trejo <mgrubentrejo@linkedin.com>

Reviewers: Gwen Shapira

Closes #6523 from mgrubent/patch-1

Commit:
fix compile error for example (#6526)

Reviewers: Prashant Sabnekar, Bill Bejeck <bbejeck@gmail.com>

viktorsomogyi committed:
KAFKA-8030: Fix flaky tests in TopicCommandWithAdminClientTest

This change adds waits for metadata updates after killing the broker in order to make the tests more stable.

Author: Viktor Somogyi-Vass <viktorsomogyi@gmail.com>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

Closes #6505 from viktorsomogyi/flaky-min-isr-test

londoncalling committed:
KAFKA-8181: Removed Avro topic from TOC on kafka (#6529)

Removed TOC entry in Streams Developer Guide for Avro, since we have no content for this

PR on kafka-site: apache/kafka-site#195

Reviewers: Guozhang Wang <wangguoz@gmail.com>

stanislavkozlovski committed:
KAFKA-8183: Add retries to WorkerUtils#verifyTopics (#6532)

Reviewers: Colin P. McCabe <cmccabe@apache.org>

guozhangwang committed:
KAFKA-7190: KIP-443; Remove streams overrides on repartition topics (#6511)

* remove streams overrides on segment.ms and segment.index.bytes

* kip comments

guozhangwang committed:
HOTFIX: Update unit test for KIP-443

adoroszlai committed:
KAFKA-8126: Flaky Test org.apache.kafka.connect.runtime.WorkerTest.testAddRemoveTask (#6475)

Changed the WorkerTest to use a mock Executor.

Author: Attila Doroszlai <adoroszlai@apache.org>
Reviewer: Randall Hauch <rhauch@gmail.com>

vvcephei committed:
KAFKA-6399: Remove Streams max.poll.interval override (#6509)

Since we now call poll during restore, we can decrease the timeout
to a reasonable value, which should help Streams make progress if
threads get stuck.

Reviewers: Guozhang Wang <wangguoz@gmail.com>,  Bill Bejeck <bbejeck@gmail.com>

mimaison committed:
KAFKA-8090: Use automatic RPC generation in ControlledShutdown

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

Closes #6423 from mimaison/controlled-shutdown

cmccabe committed:
MINOR: fix throttling and status in ConnectionStressWorker

Each separate thread should have its own throttle, so that it can sleep
for an appropriate amount of time when needed.

ConnectionStressWorker should avoid recalculating the status after
shutting down the runnables.  Otherwise, if one runnable is slow to
stop, it will skew the average down in a way that doesn't reflect
reality.  This change moves the status calculation into a separate
periodic runnable that gets shut down cleanly before the other ones.

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Gwen Shapira, Stanislav Kozlovski

Closes #6533 from cmccabe/fix_connection_stress_worker

KevinLiLu committed:
KAFKA-7904; Add AtMinIsr partition metric and TopicCommand option (KIP-427)

- Add `AtMinIsrPartitionCount` metric to `ReplicaManager`
- Add `AtMinIsr` metric to `Partition`
- Add `--at-min-isr-partitions` describe `TopicCommand` option

https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=103089398

Author: Kevin Lu <lu.kevin@berkeley.edu>
Author: lu.kevin@berkeley.edu <kelu@paypal.com>

Reviewers: Gwen Shapira

Closes #6421 from KevinLiLu/KAFKA-7904

mimaison committed:
MINOR: Fixed a few warning in core and connects (#6545)

- var -> val
- unused imports
- Javadoc fix

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

rajinisivaram committed:
KAFKA-8190; Don't update keystore modification time during validation (#6539)

Ensure that modification time is checked against the file used to create the SSLContext that is in-use so that SSLContext is updated whenever file is modified and a config update request is received.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

guozhangwang committed:
MINOR: Tighten up metadata upgrade test (#6531)

Reviewers: Bill Bejeck <bbejeck@gmail.com>

stanislavkozlovski committed:
KAFKA-7893; Refactor ConsumerBounceTest to reuse functionality from BaseConsumerTest (#6238)

This PR should help address the flakiness in the ConsumerBounceTest#testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup test (https://issues.apache.org/jira/browse/KAFKA-7965). I tested this locally and have verified it significantly reduces flakiness - 25/25 tests now pass. Running the test 25 times in trunk, I'd get `18/25` passes.

It does so by reusing the less-flaky consumer integration testing functionality inside `BaseConsumerTest`. Most notably, the test now makes use of the `ConsumerAssignmentPoller` class  - each consumer now polls non-stop rather than the more batch-oriented polling we had in `ConsumerBounceTest#waitForRebalance()`.

Reviewers: Jason Gustafson <jason@confluent.io>

cmccabe committed:
KAFKA-8168; Add a generated ApiMessageType class

Author: Colin P. Mccabe <cmccabe@confluent.io>

Reviewers: Gwen Shapira

Closes #6510 from cmccabe/KAFKA-8168

hachikuji committed:
MINOR: Move common consumer tests out of abstract consumer class (#6548)

ConsumerBounceTest redundantly executes a couple test cases which were included in the abstract class `BaseConsumerTest`. We should try to keep a cleaner separation of testing logic and utility logic so that this does not happen (the build time is long enough without doing unnecessary work). This PR moves the cluster initialization and consumer utilities out of BaseConsumerTest and into a new class AbstractConsumerTest. We then let ConsumerBounceTest extend AbstractConsumerTest.

Reviewers: Guozhang Wang <wangguoz@gmail.com>

C0urante committed:
KAFKA-8058: Fix ConnectClusterStateImpl.connectors() method (#6384)

Fixed the ConnectClusterStateImpl.connectors() method and throw an exception on timeout. Added unit test.

Author: Chris Egerton <chrise@confluent.io>

Reviewers: Magesh Nandakumar <magesh.n.kumar@gmail.com>, Robert Yokota <rayokota@gmail.com>, Arjun Satish <wicknicks@users.noreply.github.com>, Konstantine Karantasis <konstantine@confluent.io>, Randall Hauch <rhauch@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #6384 from C0urante:kafka-8058

kkonstantine committed:
KAFKA-8013; Avoid underflow when reading a Struct from a partially correct buffer (#6340)

Protocol compatibility can be facilitated if a Struct, that has been defined as an extension of a previous Struct by adding fields at the end of the older version, can read a message of an older version by ignoring the absence of the missing new fields. Reading the missing fields should be allowed by the definition of these fields (they have to be nullable) when supported by the schema.

Reviewers: David Arthur <mumrah@gmail.com>, Randall Hauch <rhauch@gmail.com>, Jason Gustafson <jason@confluent.io>

cadonna committed:
MINOR: Correct KStream documentation (#6552)

Reviewers: Bill Bejeck <bill@confluent.io>, Matthias J. Sax <matthias@confluent.io>

highluck committed:
MINOR: Remove redundant access specifiers from metrics interfaces (#6527)

Reviewers: Sönke Liebau <soenke.liebau@opencore.com>, Ismael Juma <ismael@juma.me.uk>

rajinisivaram committed:
MINOR: Add security considerations for remote JMX in Kafka docs (#6544)

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

ableegoldman committed:
[MINOR] Guard against crashing on invalid key range queries (#6521)

Due to KAFKA-8159, Streams will throw an unchecked exception when a caching layer or in-memory underlying store is queried over a range of keys from negative to positive. We should add a check for this and log it then return an empty iterator (as the RocksDB stores happen to do) rather than crash

Reviewers: Bruno Cadonna <bruno@confluent.io> Bill Bejeck <bbejeck@gmail.com>

hachikuji committed:
MINOR: Use generated InitProducerId RPC (#6538)

This patch updates the InitProducerId request API to use the generated sources. It also fixes a small bug in the DescribeAclsRequest class where we were using the wrong api key.

Reviewers: Mickael Maison <mickael.maison@gmail.com>, Colin McCabe <cmccabe@apache.org>

cmccabe committed:
MINOR: ConnectionStressWorker: add missing executor shutdown (#6558)

hachikuji committed:
MINOR: Remove SubscriptionState.Listener and replace with assignmentId tracking (#6559)

We have not had great experience with listeners. They make the code harder to understand because they result in indirectly maintained circular dependencies. Often this leads to tricky deadlocks when we try to introduce locking. We were able to remove the Metadata listener in KAFKA-7831. Here we do the same for the listener in SubscriptionState.

Reviewers: Viktor Somogyi-Vass <viktorsomogyi@gmail.com>, Rajini Sivaram <rajinisivaram@googlemail.com>

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImplLogAndContinueTest.java:
```diff
@@ -0,0 +1,129 @@
+import static org.apache.kafka.common.utils.Utils.mkMap;
+import static org.junit.Assert.assertEquals;
+
+public class GlobalStateManagerImplLogAndContinueTest extends GlobalStateManagerImplTest {
```
What's the utility of extending another test instead of just either making a new test or adding stuff to the other test?

vvcephei replied:
I see now that there's been some conversation about this. I'll just mention that I'd be in favor of seeking some other solution to the testing problem.

pkleindl replied:
I'm open for alternatives, because I don't like the solution, but I would have had to change a lot in the existing GlobalStateManagerImplTest to allow parametrization of the StreamsConfig.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:
```diff
@@ -275,11 +298,16 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
                     final ConsumerRecords<byte[], byte[]> records = globalConsumer.poll(pollTime);
                     final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>();
                     for (final ConsumerRecord<byte[], byte[]> record : records.records(topicPartition)) {
+                        if (recordDeserializer != null && recordDeserializer.deserialize(processorContext, record) == null) {
```
What are the semantics of the deserializer returning null? It seems like it should be a tombstone, but we're instead skipping the store update?

pkleindl replied:
> It seems like global stores can only be populated from input topics, is that right? In that case, they probably never need a changelog and just instead always populate and restore from the input. If that's true, then why do global stores need a concept of "restore" at all? It seems at first brush like populate and restore are functionally equivalent.

This problem is not specific to global stores, I just did this part of the implementation first and would extend it to the local stores afterwards.
The point is that any unserializable record in the source topic can be handled during normal operations but will be loaded in the store upon restore, potentially causing a problem when accessing the record from the store.
And without this kind of protection there is not a lot you can do except try to clean up the topic itself or try to catch deserialization problems anywhere the store is accessed.

> Maybe the restore path is more efficient for some reason, like it polls in larger batches or something?

It would be a lot more efficient if not every record had to be inspected, although I didn't change the batch loading (hopefully). I only try to deserialize the record and skip it when it fails.

pkleindl replied:
When using the LogAndContinueExceptionHandler the deserialize returns null for the whole ConsumerRecord.
The tombstone should have a key set and value null, right?
I will try to add a test for the tombstone case, just to be sure.

londoncalling committed:
KAFKA-8213 - Fix typo in Streams Dev Guide (#6574)

Reviewers: Jim Galasyn <jim.galasyn@confluent.io>, Joel Hamill <joel@confluent.io>, Matthias J. Sax <matthias@confluent.io>

londoncalling committed:
KAFKA-8212 DOCS (kafka) - Fix Maven artifacts table from cutting off text (#6576)

Reviewer: Matthias J. Sax <matthias@confluent.io>

bbejeck committed:
KAFKA-8209: Wrong link for KStreams DSL in core concepts doc (#6564)

Reviewers Matthias J. Sax <mjsax@apache.org>, Michael Drogalis <michael.drogalis@confluent.io>, Victoria Bialas <vicky@confluent.io>

bbejeck committed:
KAFKA-8210: Fix link for streams table duality (#6573)

Reviewers: Victoria Bialas <vicky@confluent.io>

bbejeck committed:
KAFKA-8208: Change paper link directly to ASM (#6572)

Reviewers: Matthias J. Sax <mjsax@apache.org>, Victoria Bialas <vicky@confluent.io>

rajinisivaram committed:
KAFKA-8232; Test topic delete completion rather than intermediate state (#6581)

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>

vvcephei commented:
Hi @pkleindl ,

Sorry for the long delay. I'm coming up to speed on this, and I have a high-level question, and please bear in mind that I don't know much about global stores to begin with... 

It seems like global stores can only be populated from input topics, is that right? In that case, they probably never need a changelog and just instead always populate and restore from the input. If that's true, then why do global stores need a concept of "restore" at all? It seems at first brush like populate and restore are functionally equivalent.

Maybe the restore path is more efficient for some reason, like it polls in larger batches or something?

I'm asking all this because it seems like there's some risk and complexity involved in maintaining two separate paths to populate the store. This is already the case with our pre-existing restore path, which is the ultimate cause of the bug you're fixing.

Thanks! I had a couple of small remarks, which I'll also send, but I wanted to ask for clarification on the big picture.

soenkeliebau committed:
MINOR: Mention in configuration of broker setting log.retention.ms that -1 disables retention by time (#6464)

Includes an update to the relevant configuration doc.

pkleindl committed:
KAFKA-8200: Changed Queue to Deque for FIFO guarantee on outputRecordsByTopic

pkleindl committed:
Merge remote-tracking branch 'origin/KAFKA-8037' into KAFKA-8037

pkleindl commented:
@vvcephei @bbejeck Sorry for asking directly, but I think I messed up when updating my branch.
I didn't mean to include all other changed from trunk here.
Do you know if there is a simple way to remove those changes? Or should I rather start over again in a new PR?

vvcephei commented:
@pkleindl , no worries. You can use "rebase" to get rid of those commits. Just look in the commit history to find the commit that this whole branch is based on, and run `git rebase -i ${THAT_COMMIT}`. It'll give you a text file listing all the commits in the branch, and you just delete any line you don't want included. Afterwards, you do a force-push to your PR branch.

Tread with caution, obviously. You can create a new branch or tag at the head of this branch before you do the rebase, so you can "undo" if things go sideways, `git checkout KAFKA-8037; git checkout -b KAFKA-8037-backup; git checkout KAFKA-8037; git rebase -i ...`. Then, if the rebase messes everything up, you just `git checkout KAFKA-8037-backup; git branch -D KAFKA-8037; git checkout -b KAFKA-8037` to restore.

vvcephei commented:
Moving this back into the main thread:

>> It seems like global stores can only be populated from input topics, is that right? In that case, they probably never need a changelog and just instead always populate and restore from the input. If that's true, then why do global stores need a concept of "restore" at all? It seems at first brush like populate and restore are functionally equivalent.

>This problem is not specific to global stores, I just did this part of the implementation first and would extend it to the local stores afterwards.
The point is that any unserializable record in the source topic can be handled during normal operations but will be loaded in the store upon restore, potentially causing a problem when accessing the record from the store.
And without this kind of protection there is not a lot you can do except try to clean up the topic itself or try to catch deserialization problems anywhere the store is accessed.

So, for local stores, there are two modes... In general, the store has a changelog, and we restore from the changelog. Since we only write to the changelog *after* record processing, we know it only contains valid data. In special cases, as an optimization, we don't create a changelog and instead restore from the input topic. In that latter case, it is indeed the same problem as what you're addressing here.

My question is, in the cases where we're "restoring" from the input topic, rather than from the changelog, why bother with the "restore" code path at all, why not just process the input topic normally? The RecordQueue would ensure that we read all the older records from this store's input before processing newer records from other topics anyway. This effectively means that we'd "restore" the state of the store in question before trying to use it for joins, etc., which I think is what really matters in the end.

pkleindl commented:
@vvcephei Thanks for the help, will try the rebase tomorrow

As for your suggestion, the idea sounds very good to me and more elegant than my approach.
Some possible restrictions/drawbacks:
- If this case is handled by normal processing, wouldn't this be quite limited in speed when the normal consumer is used?
- This would possibly limit the performance for this kind of "restore" as the restore consumer can be tuned differently than the normal consumer
- I am not sure if this is equally applicable for the global stores because their restore is different from the local ones and they have to be completely restored before any other processing (including restoration of local stores) can start

Should be take this back to the mailing list or try to get some opinions here?
We aren't really talking about implementation "details" at the moment ;-)

vvcephei commented:
Hey @pkleindl ,

Sorry for the silence. I'm not sure if I saw that mailing list message; just re-starting it here.

I do think that performance is a concern for "restoring" a store from its input instead of from a changelog, but it seems like this could have a pretty straightforward solution of just configuring the consumer and store for bulk during the initial "bulk" phase. It's possible, but not clear to be whether _also_ skipping some of the processing/handling logic also meaningfully affects performance or not.

ableegoldman commented:
Just to chime in here, I think we actually don't have a separate restore consumer for global stores -- this makes sense for the case of local stores, where the "normal" consumers use group subscription to subscribe to source topics while the restore consumers are assigned directly to specific changelog topics.
However for global stores we are restoring and "processing" from the same input topic, so the only real difference is the code path. Which as John has pointed out should not be meaningfully different between the "restore" and "initial loading" cases.

pkleindl commented:
@vvcephei  @ableegoldman 
Thanks for your feedback. It is correct that the global stores have their own logic now, although it would be very favorable if this could be changed to allow parallel restore/initialization at least.
Maybe this problem can be reduced to: Any store based on a source topic can run into this problem during initialization if the records are sent straight to the store without deserialization.
I am not sure if I understand the above comment correctly, but using the normale consumer instead of the restore consumer would run into other problems when building up the stores, right?

Unless everyone is using Avro I think more people will run into this kind of problem which is pretty nasty and workarounds are not really easy.

But I can give it another try if we find solution which is safe and fast (enough).

vvcephei commented:
Thanks, @pkleindl !

I _do_ agree it's a nasty problem, and to be honest, I've always been a little hesitant about using the input instead of a changelog topic. It's definitely a disk/network saver, but it also seems like a big risk to me.

Looking forward to see what you come up with!

lkokhreidze commented:
Sorry for reviving old PR, but I have one high-level question about this, sorry if i am missing something, but if the aim is to stop populating records that can't be deserialized into the global store, why can't we have something like this in the `GlobalStateManagerImpl#restoreState`

```
            while (offset < highWatermark) {
                try {
                    final ConsumerRecords<byte[], byte[]> records = globalConsumer.poll(pollTime);
                    final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>();
                    for (final ConsumerRecord<byte[], byte[]> record : records.records(topicPartition)) {
                        if (record.key() != null) {
                            final String topic = topicPartition.topic();

                            final SourceNode source = topology.source(topic);
                            // try to deserialize key and value, if not use provided 
                            // deserelization exception handling strategy to either fail topology or continue
                            final Object key = source.deserializeKey(topic, record.headers(), record.key());
                            final Object value = source.deserializeValue(topic, record.headers(), record.value());

                            restoreRecords.add(recordConverter.convert(record));

                        }
                    }
                   ...
                } catch (final InvalidOffsetException recoverableException) {
                   ...
                }
            }
```

During restoration we can try to deserialize key and value with the provided deserializers, if not successful we can use provided `DeserializationExceptionHandler` to either continue or fail altogether?

pkleindl commented:
Hi, that was my basic idea, see https://github.com/apache/kafka/pull/6398/commits/8ea4d944bb5389d2014e4f3b4a617255a7179843#diff-e870eab2dedeee93442ba87dc12609a4
But I did not have the time to continue working on it in the last months, so the progress stalled.
This was meant as a basis to use the process during restoration too, but all this would need a bit of discussion because of possible performance impacts.

guozhangwang commented:
@lkokhreidze I looked at another JIRA ticket related to your observation and left some comments there (I think they are relevant :P): https://issues.apache.org/jira/browse/KAFKA-7663 LMK what do you think.

pkleindl commented:
I'll close this PR as the differences are too big after this time.
A new PR will be opened
